bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/LocalFolderWatcherHandler.java

```diff
@@ -0,0 +1,164 @@
+import org.openhab.binding.folderwatcher.internal.config.LocalFolderWatcherConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link LocalFolderWatcherHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Alexandr Salamatov - Initial contribution
+ */
+@NonNullByDefault
+public class LocalFolderWatcherHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(LocalFolderWatcherHandler.class);
+    private LocalFolderWatcherConfiguration config = new LocalFolderWatcherConfiguration();
+    private File currentLocalListingFile = new File(ConfigConstants.getUserDataFolder() + File.separator
+            + "FolderWatcher" + File.separator + thing.getUID().getAsString().replace(':', '_') + ".data");
+    private @Nullable ScheduledFuture<?> executionJob;
+    private List<String> currentLocalListing = new ArrayList<>();
+    private List<String> previousLocalListing = new ArrayList<>();
+
+    public LocalFolderWatcherHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Channel {} triggered with command {}", channelUID.getId(), command);
+        if (command instanceof RefreshType) {
+            refreshFolderInformation();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(LocalFolderWatcherConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+
+        if (!Files.isDirectory(Paths.get(config.localDir))) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Local directory is not valid");
+            return;
+        }
+        try {
+            previousLocalListing = WatcherCommon.initStorage(currentLocalListingFile);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            logger.debug("Can't write file {}: {}", currentLocalListingFile, e.getMessage());
+            return;
+        }
+
+        if (config.pollIntervalLocal > 0) {
+            updateStatus(ThingStatus.ONLINE);
+            executionJob = scheduler.scheduleWithFixedDelay(this::refreshFolderInformation, config.pollIntervalLocal,
+                    config.pollIntervalLocal, TimeUnit.SECONDS);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Polling interval can't be null or negative");
+            return;
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> executionJob = this.executionJob;
+        if (executionJob != null) {
+            executionJob.cancel(true);
+        }
+    }
+
+    private void refreshFolderInformation() {
+        final String rootDir = config.localDir;
+        try {
+            currentLocalListing.clear();
```

Comment:
To avoid any concurrency issues I would suggest just creating a new list each time instead of trying to reuse the old instance.

Reply:
Fixed